pkg/controller/perconaservermongodb/smart.go

```diff
@@ -116,11 +140,151 @@ func (r *ReconcilePerconaServerMongoDB) smartUpdate(cr *api.PerconaServerMongoDB
 		return fmt.Errorf("failed to apply changes: %v", err)
 	}
 
-	log.Info("smart update finished")
+	err = r.startBalancerIfNeeded(cr, username, password)
+	if err != nil {
+		return errors.Wrap(err, "failed to start balancer")
+	}
+
+	log.Info("smart update finished for statefulset", "statefulset", sfs.Name)
+
+	return nil
+}
+
+func stopBalancerIfNeeded(cr *api.PerconaServerMongoDB, username, password string) error {
+	if !cr.Spec.Sharding.Enabled {
+		return nil
+	}
+
+	mongosSession, err := mongosConn(cr, username, password)
+	if err != nil {
+		return errors.Wrap(err, "failed to get mongos connection")
+	}
+
+	defer func() {
+		err := mongosSession.Disconnect(context.TODO())
+		if err != nil {
+			log.Error(err, "failed to close mongos connection")
+		}
+	}()
+
+	err = mongo.StopBalancer(context.TODO(), mongosSession)
+	if err != nil {
+		return errors.Wrap(err, "failed to stop balancer")
+	}
+
+	log.Info("balancer stopped")
 
 	return nil
 }
 
+func (r *ReconcilePerconaServerMongoDB) isAllSfsUpToDate(cr *api.PerconaServerMongoDB) (bool, error) {
+	sfsList := appsv1.StatefulSetList{}
+	if err := r.client.List(context.TODO(), &sfsList,
+		&client.ListOptions{
+			Namespace: cr.Namespace,
+			LabelSelector: labels.SelectorFromSet(map[string]string{
+				"app.kubernetes.io/instance": cr.Name,
+			}),
+		},
+	); err != nil {
+		return false, errors.Wrap(err, "failed to get statefulset list")
+	}
+
+	for _, s := range sfsList.Items {
+		if s.Status.UpdatedReplicas < s.Status.Replicas {
+			return false, nil
+		}
+	}
+
+	return true, nil
+}
+
+func (r *ReconcilePerconaServerMongoDB) startBalancerIfNeeded(cr *api.PerconaServerMongoDB, username, password string) error {
+	if !cr.Spec.Sharding.Enabled {
+		return nil
+	}
+
+	uptodate, err := r.isAllSfsUpToDate(cr)
+	if err != nil {
+		return errors.Wrap(err, "failed to chaeck if all sfs are up to date")
+	}
+
+	if !uptodate {
+		return nil
+	}
+
+	mongosPods := corev1.PodList{}
+	err = r.client.List(context.TODO(),
+		&mongosPods,
+		&client.ListOptions{
+			Namespace: cr.Namespace,
+			LabelSelector: labels.SelectorFromSet(map[string]string{
+				"app.kubernetes.io/name":       "percona-server-mongodb",
+				"app.kubernetes.io/instance":   cr.Name,
+				"app.kubernetes.io/managed-by": "percona-server-mongodb-operator",
+				"app.kubernetes.io/part-of":    "percona-server-mongodb",
+				"app.kubernetes.io/component":  "mongos",
+			}),
+		},
+	)
```

Comment:
can we check Deployment here? and check state identically to StatefulSet?
```
if s.Status.UpdatedReplicas < s.Status.Replicas {
```

Reply:
https://github.com/percona/percona-server-mongodb-operator/pull/449/commits/c9397a3f456df78e2d9b7ef67e8b5caf6212a2c8